src/wayland/input_method/mod.rs

```diff
@@ -0,0 +1,364 @@
+                f.flush()?;
+                keyboard.keymap(
+                    KeymapFormat::XkbV1,
+                    f.as_raw_fd(),
+                    keyboard_state.keymap.as_bytes().len() as u32,
+                );
+                Ok(())
+            })
+            .expect("File not working!");
+    }
+
+    // return true if modifier state has changed
+    fn key_input(&mut self, keycode: u32, keystate: KeyState) -> bool {
+        // track pressed keys as xkbcommon does not seem to expose it :(
+        let direction = match keystate {
+            KeyState::Pressed => xkb::KeyDirection::Down,
+            KeyState::Released => xkb::KeyDirection::Up,
+        };
+
+        // update state
+        // Offset the keycode by 8, as the evdev XKB rules reflect X's
+        // broken keycode system, which starts at 8.
+        let state = &mut self.keyboard_state.as_mut().unwrap().xkbstate;
+        let state_components = state.update_key(keycode + 8, direction);
+
+        state_components != 0
+    }
+
+    fn serialize_modifiers(&self) -> (u32, u32, u32, u32) {
+        let state = &self.keyboard_state.as_ref().unwrap().xkbstate;
+        let mods_depressed = state.serialize_mods(xkb::STATE_MODS_DEPRESSED);
+        let mods_latched = state.serialize_mods(xkb::STATE_MODS_LATCHED);
+        let mods_locked = state.serialize_mods(xkb::STATE_MODS_LOCKED);
+        let layout_locked = state.serialize_layout(xkb::STATE_LAYOUT_LOCKED);
+
+        (mods_depressed, mods_latched, mods_locked, layout_locked)
+    }
+
+    fn input(
+        &self,
+        keycode: u32,
+        state: KeyState,
+        serial: Serial,
+        time: u32,
+        update_modifiers: bool, //TODO: Add modifier handling from closure
+                                //modifiers:ModifiersState
+    ) {
+        let keyboard = self.keyboard.as_ref().unwrap();
+        let wl_state = match state {
+            KeyState::Pressed => WlKeyState::Pressed,
+            KeyState::Released => WlKeyState::Released,
+        };
+        keyboard.key(serial.into(), time, keycode, wl_state);
+
+        if update_modifiers {
+            let (dep, la, lo, gr) = self.serialize_modifiers();
+            keyboard.modifiers(serial.into(), dep, la, lo, gr);
+        }
+    }
+}
+
+fn xkb_handler(xkb_config: XkbConfig<'_>) -> (String, xkb::State) {
+    let context = xkb::Context::new(xkb::CONTEXT_NO_FLAGS);
+    let keymap = xkb::Keymap::new_from_names(
+        &context,
+        &xkb_config.rules,
+        &xkb_config.model,
+        &xkb_config.layout,
+        &xkb_config.variant,
+        xkb_config.options,
+        xkb::KEYMAP_COMPILE_NO_FLAGS,
+    )
+    .unwrap();
+    let state = xkb::State::new(&keymap);
+    (keymap.get_as_string(xkb::KEYMAP_FORMAT_TEXT_V1), state)
+}
+
+impl fmt::Debug for InputMethod {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("InputMethod").finish()
+    }
+}
```

Comment:
Could we move this manual impl to `KeyboardState` (as it is the reason for manual impl in the first place), and this way we will also preserver `rate`, `delay`, `keymap` in the debug log.

Reply:
I'll fix it when I get around too it, thanks

Reply:
Done